src/services/gcs/dir_stream.rs

```diff
@@ -0,0 +1,307 @@
+
+                    let de = DirEntry::new(
+                        backend.clone(),
+                        ObjectMode::FILE,
+                        &backend.get_rel_path(&object.name),
+                    );
+
+                    log::debug!(
+                        "dir object {} got entry, mode: {}, path: {}",
+                        &self.path,
+                        de.mode(),
+                        de.path()
+                    );
+                    return Poll::Ready(Some(Ok(de)));
+                }
+
+                // end of asynchronous iteration
+                if self.done {
+                    log::debug!("object {} list done", &self.path);
+                    return Poll::Ready(None);
+                }
+
+                self.state = State::Standby;
+                self.poll_next(cx)
+            }
+        }
+    }
+}
+
+/// Response JSON from GCS list objects API.
+///
+/// refer to https://cloud.google.com/storage/docs/json_api/v1/objects/list for details
+#[derive(Default, Debug, Deserialize)]
+#[serde(default, rename_all = "camelCase")]
+struct Output {
+    /// The kind of item this is.
+    /// For lists of objects, this is always "storage#objects"
+    kind: String,
+    /// The continuation token.
+    ///
+    /// If this is the last page of results, then no continuation token is returned.
+    next_page_token: Option<String>,
+    /// Object name prefixes for objects that matched the listing request
+    /// but were excluded from [items] because of a delimiter.
+    prefixes: Vec<String>,
+    /// The list of objects, ordered lexicographically by name.
+    items: Vec<OutputContent>,
+}
+
+#[derive(Default, Debug, Eq, PartialEq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct OutputContent {
+    name: String,
+    size: String,
+}
+
+#[cfg(test)]
+mod ds_test {
+    use crate::services::gcs::dir_stream::Output;
+
+    #[test]
+    fn test_de_list() {
+        let names = vec![
+            "/home/datafuselabs/",
+            "/home/datafuselabs/databend.txt",
+            "/home/datafuselabs/cache-server.md",
+        ];
+        let sizes = vec![10000, 5000, 5000];
+        let objects = names
+            .iter()
+            .zip(sizes.iter())
+            .map(|(&name, size)| mock_object(name.to_string(), *size))
+            .collect();
+        let list = mock_list(objects);
+
+        let output_list = serde_json::de::from_str::<Output>(list.as_str());
+        assert!(output_list.is_ok());
+        let output_list = output_list.unwrap();
+        assert_eq!(output_list.items.len(), 3);
+        for (idx, item) in output_list.items.iter().enumerate() {
+            assert_eq!(item.name, names[idx]);
+            let size_res = item.size.as_str().parse();
+            assert!(size_res.is_ok());
+            let size: u64 = size_res.unwrap();
+            assert_eq!(size, sizes[idx]);
+        }
+    }
+
+    fn mock_object(name: String, size: u64) -> String {
```

Comment:
Please don't introduce new logic here (we don't know if mock_object is correct). Just use the raw content from API response.